src/S3/PostObjectV4.php

```diff
@@ -0,0 +1,191 @@
+        ];
+
+        // setup basic policy
+        $expiration = TimestampShape::format($expiration, 'iso8601');
+
+        $policy = [
+            'expiration' => $expiration,
+            'conditions' => [
+                ['X-Amz-Algorithm' => "AWS4-HMAC-SHA256"],
+            ],
+        ];
+        foreach ($options as $option) {
+            $policy['conditions'][] = $option;
+        }
+        // setup basic formInputs
+        $this->formInputs = $formInputs + ['key' => '${filename}'];
+
+        // finalize policy and signature
+        $credentials = $this->client->getCredentials()->wait();
+        $this->formInputs += $this->getPolicyAndSignature(
+            $credentials,
+            $policy
+        );
+    }
+
+    /**
+     * Gets the S3 client.
+     *
+     * @return S3ClientInterface
+     */
+    public function getClient()
+    {
+        return $this->client;
+    }
+
+    /**
+     * Gets the bucket name.
+     *
+     * @return string
+     */
+    public function getBucket()
+    {
+        return $this->bucket;
+    }
+
+    /**
+     * Gets the form attributes as an array.
+     *
+     * @return array
+     */
+    public function getFormAttributes()
+    {
+        return $this->formAttributes;
+    }
+
+    /**
+     * Set a form attribute.
+     *
+     * @param string $attribute Form attribute to set.
+     * @param string $value     Value to set.
+     */
+    public function setFormAttribute($attribute, $value)
+    {
+        $this->formAttributes[$attribute] = $value;
+    }
+
+    /**
+     * Gets the form inputs as an array.
+     *
+     * @return array
+     */
+    public function getFormInputs()
+    {
+        return $this->formInputs;
+    }
+
+    /**
+     * Set a form input.
+     *
+     * @param string $field Field name to set
+     * @param string $value Value to set.
+     */
+    public function setFormInput($field, $value)
+    {
+        $this->formInputs[$field] = $value;
+    }
+
+    private function generateUri()
+    {
+        $uri = new Uri($this->client->getEndpoint());
+
+        if ($uri->getScheme() === 'https'
+            && strpos($this->bucket, '.') !== false
+        ) {
+            // Use path-style URLs
+            $uri = $uri->withPath($this->bucket);
+        } else {
+            // Use virtual-style URLs
+            $uri = $uri->withHost($this->bucket . '.' . $uri->getHost());
+        }
+
+        return (string) $uri;
+    }
+
+    protected function getPolicyAndSignature(
+        CredentialsInterface $credentials,
+        $policy
+    ){
+        $ldt = gmdate(SignatureV4::ISO8601_BASIC);
+        $sdt = substr($ldt, 0, 8);
+        $policy['conditions'][] = ['X-Amz-Date' => $ldt];
+
+        $region = $this->client->getRegion();
+        $scope = $this->createScope($sdt, $region, 's3');
+        $creds = "{$credentials->getAccessKeyId()}/$scope";
+        $policy['conditions'][] = ['X-Amz-Credential' => $creds];
```

Comment:
For multiple array additions, you can use the union operator (`+`) instead of pushing each one on individually:
```php
$policy['conditions'] += [
    'X-Amz-Date' => $ldt,
    'X-Amz-Credential' => $creds,
    'X-Amz-Algorithm' => 'AWS-HMAC-SHA256',
];
```